src/Stratis.Bitcoin.Features.Consensus/Rules/CommonRules/PosTransactionRelativeLocktimeAndSignatureOperationCostRule.cs

```diff
@@ -0,0 +1,151 @@
+﻿using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Base.Deployments;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.Consensus.Rules.CommonRules
+{
+    /// <summary>
+    /// Checks each transaction conforms to BIP68 Final (Time and Blockheight checks) - see https://github.com/bitcoin/bips/blob/master/bip-0068.mediawiki
+    /// And checks signature operation costs
+    /// Then updates the coinview with each transaction.
+    /// With the addition of coinview update and maturity variations for Proof of Stake.
+    /// </summary>
+    [ExecutionRule]
+    public class PosTransactionRelativeLocktimeAndSignatureOperationCostRule : PowTransactionRelativeLocktimeAndSignatureOperationCostRule
+    {
+        /// <summary>Consensus options.</summary>
+        private PosConsensusOptions consensusOptions;
+
+        public override void Initialize()
+        {
+            this.consensusOptions = this.Parent.Network.Consensus.Option<PosConsensusOptions>();
+        }
+
+        public override Task RunAsync(RuleContext context)
+        {
+            Block block = context.BlockValidationContext.Block;
+            ChainedBlock index2 = context.BlockValidationContext.ChainedBlock;
+            DeploymentFlags flags = context.Flags;
+            UnspentOutputSet view = context.Set;
+
+            this.Parent.PerformanceCounter.AddProcessedBlocks(1);
+
+            long sigOpsCost = 0;
+
+            var fees = Money.Zero;
+
+            context.CheckInputs = new List<Task<bool>>();
+            for (int txIndex = 0; txIndex < block.Transactions.Count; txIndex++)
```

Comment:
The index is just to get the Transaction from the block (on line 44) - could use foreach-loop.